pkg/agent/svid/store/service.go

```diff
@@ -0,0 +1,270 @@
+package store
+
+import (
+	"context"
+	"crypto/x509"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/andres-erbsen/clock"
+	"github.com/sirupsen/logrus"
+	"github.com/spiffe/go-spiffe/v2/spiffeid"
+	"github.com/spiffe/spire/pkg/agent/catalog"
+	"github.com/spiffe/spire/pkg/agent/manager/storecache"
+	"github.com/spiffe/spire/pkg/agent/plugin/svidstore"
+	"github.com/spiffe/spire/pkg/common/telemetry"
+	telemetry_store "github.com/spiffe/spire/pkg/common/telemetry/agent/store"
+	"github.com/spiffe/spire/pkg/common/util"
+	"github.com/spiffe/spire/proto/spire/common"
+)
+
+const (
+	defaultInterval = 5 * time.Second
+)
+
+type Cache interface {
+	// ReadyToStore is a list of store cache records that are ready to be stored on specific SVID Store
+	ReadyToStore() []*storecache.Record
+	// HandledRecord sets a revision to record on cache
+	HandledRecord(entry *common.RegistrationEntry, revision int64)
+}
+
+type Config struct {
+	Clk         clock.Clock
+	Log         logrus.FieldLogger
+	TrustDomain spiffeid.TrustDomain
+	Cache       Cache
+	Catalog     catalog.Catalog
+	Metrics     telemetry.Metrics
+}
+
+type SVIDStoreService struct {
+	clk clock.Clock
+	log logrus.FieldLogger
+	// trustDomain is the trust domain of the agent
+	trustDomain spiffeid.TrustDomain
+	// cache is the store cache
+	cache Cache
+	// svidStores is the allowed list of SVID Stores configured on agent
+	svidStores map[string]svidstore.SVIDStore
+	metrics    telemetry.Metrics
+
+	hooks struct {
+		// test hook used to verify if a cycle finished
+		storeFinished chan struct{}
+	}
+}
+
+func New(c *Config) *SVIDStoreService {
+	clk := c.Clk
+	if clk == nil {
+		clk = clock.New()
+	}
+
+	svidStores := make(map[string]svidstore.SVIDStore)
+	for _, store := range c.Catalog.GetSVIDStores() {
+		svidStores[store.Name()] = store
+	}
+
+	return &SVIDStoreService{
+		cache:       c.Cache,
+		clk:         clk,
+		log:         c.Log,
+		metrics:     c.Metrics,
+		trustDomain: c.TrustDomain,
+		svidStores:  svidStores,
+	}
+}
+
+// SetStoreFinishedHook used for testing only
+func (s *SVIDStoreService) SetStoreFinishedHook(storeFinished chan struct{}) {
+	s.hooks.storeFinished = storeFinished
+}
+
+// Run starts SVID Store service
+func (s *SVIDStoreService) Run(ctx context.Context) error {
+	// TODO: may we run a StoreSVID when `Run` start?
+	s.processRecords(ctx)
+
+	ticker := s.clk.Ticker(defaultInterval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			s.processRecords(ctx)
+		case <-ctx.Done():
+			return nil
+		}
+	}
+}
```

Comment:
This can lead to back-to-back processing if processRecords takes any amount of time (likely). Maybe something like this?

```
timer := s.clk.Timer(defaultInterval)
defer timer.Stop()

for {
        s.processRecords(ctx)
        timer.Reset(defaultInterval)
        select {
        case <-timer.C:
        case <-ctx.Done():
            return nil
        }
}
```

This also takes care of the TODO above...